cycle detected when optimizing MIR with mir-opt-level=2

```rust
#![crate_type = "lib"]

#[derive(Clone, Debug)]
pub struct S {}

impl S {
    pub async fn g(mut c: Self) {
        Self::h(&mut c);
    }

    pub fn h(c: &mut Self) {
        let _ = S::g(c.clone());
    }
}
```

```console
$ rustc --edition=2018  -Zmir-opt-level=2 a.rs
error[E0391]: cycle detected when optimizing MIR for `<impl at a.rs:6:1: 14:2>::g::{{closure}}#0`
  --> a.rs:7:33
   |
7  |       pub async fn g(mut c: Self) {
   |  _________________________________^
8  | |         Self::h(&mut c);
9  | |     }
   | |_____^
   |
note: ...which requires optimizing MIR for `<impl at a.rs:6:1: 14:2>::h`...
  --> a.rs:11:5
   |
11 |     pub fn h(c: &mut Self) {
   |     ^^^^^^^^^^^^^^^^^^^^^^
   = note: ...which requires computing layout of `std::future::from_generator::GenFuture<[static generator@a.rs:7:33: 9:6 __arg0:S {}]>`...
   = note: ...which requires computing layout of `[static generator@a.rs:7:33: 9:6 __arg0:S {}]`...
   = note: ...which again requires optimizing MIR for `<impl at a.rs:6:1: 14:2>::g::{{closure}}#0`, completing the cycle
```

--- a/compiler/rustc_mir/src/transform/inline.rs
+++ b/compiler/rustc_mir/src/transform/inline.rs
@@ -107,8 +107,14 @@ impl Inliner<'tcx> {
                     // Avoid a cycle here by only using `optimized_mir` only if we have
                     // a lower `HirId` than the callee. This ensures that the callee will
                     // not inline us. This trick only works without incremental compilation.
-                    // So don't do it if that is enabled.
-                    if !self.tcx.dep_graph.is_fully_enabled() && self_hir_id < callee_hir_id {
+                    // So don't do it if that is enabled. Also avoid inlining into generators,
+                    // since their `optimized_mir` is used for layout computation, which can
+                    // create a cycle, even when no attempt is made to inline the function
+                    // in the other direction.
+                    if !self.tcx.dep_graph.is_fully_enabled()
+                        && self_hir_id < callee_hir_id
+                        && caller_body.generator_kind.is_none()
+                    {
                         self.tcx.optimized_mir(callsite.callee)
                     } else {
                         continue;

--- a/src/test/mir-opt/inline/inline-async.rs
+++ b/src/test/mir-opt/inline/inline-async.rs
@@ -1,0 +1,18 @@
+// Checks that inliner doesn't introduce cycles when optimizing generators.
+// The outcome of optimization is not verfied, just the absence of the cycle.
+// Regression test for #76181.
+//
+// edition:2018
+
+#![crate_type = "lib"]
+
+pub struct S;
+
+impl S {
+    pub async fn g(&mut self) {
+        self.h();
+    }
+    pub fn h(&mut self) {
+        let _ = self.g();
+    }
+}

